Step2 PR

yunbaek opened the pull request:
안녕하세요! 백경윤입니다.

ATTD 미션 step2 - 지하철 노선 기능 구현하였습니다.

- 인수 테스트간 격리를 위해 `DatabaseCleanUp` 기능을 하는 클래스를 만들어 보았습니다. [커밋](https://github.com/next-step/atdd-subway-admin/commit/5b4fd7db6228caf913cd3de25856bcf8e899afe1)
기존에 주신 힌트(?) 는 `EntityManager` 를 통해 `@Table` 어노테이션이 달린 엔티티들의 테이블 정보를 불러오고, 각 테이블을  `Trucate` 하는 방식이었는데, 실무에서 `JPA` 를 사용하다보니  `@SubSelect` 어노테이션을 통해  서브쿼리로 사용하는 경우와 좀 더 작은 Entity로 만드는 경우 등 꼭 한 테이블이 하나의 `Entity` 로만 작성되는 것이 아니더라구요 😂
그래서 다른 방법은 없을까 고민하다가 `DataSource` 의 `metaData` 를 통해 테이블 정보를 불러오고, 해당 테이블에 대하여 `Truncate` 하는 방식으로 구현 해 보았습니다. 혹시 괜찮은 방법인지, 제가 미처 생각하지 못한 오류는 없을 지 궁금합니다 🤔 일단 인수테스트 격리는 제대로 되는 것 같습니다 😄 
- 인수테스트 작성 시 검증부에서 어디까지(?) 검증을 해야 하는지 잘 모르겠습니다 😢 
예를 들어, 조회에 대한 응답에 대해 검증 시에 응답받은 모든 항목들을 검증해야 할지, 몇 개 정도는 빼먹어도...:joy: 될지 궁금합니다. 혹시 리뷰어님께서 개인적으로 정해놓은 기준 같은 것이 있으시면 조언 얻고 싶습니다!


그럼 이번에도 따끔한 피드백 많이 부탁드리겠습니다!!
항상 감사드립니다.

yunbaek committed:
docs: step2 요구사항 정리

yunbaek committed:
chore: 패키지 세분화

yunbaek committed:
test: Station 테스트 픽스처 생성

- StationGenerator
- Station 정적 팩토리 메서드 생성

yunbaek committed:
feat: Line Entity & Repository 구현

- CRUD 테스트 추가

yunbaek committed:
chore: 테스트 유틸 수정

- 패키지 변경
- post 테스트 타입 변경(Map<String, String> -> <T>)

yunbaek committed:
test: 노선 생성 인수 테스트 추가

- 공통 util 생성
- NotFound 예외 생성
- 노선 Response & CreateRequest 객체 생성

Commit:
feat: Line 생성을 위한 LineService, Test 작성

Commit:
feat: 지하철 노선 생성 API 구현

- 지하철 생성 인수 테스트 수정
- 지하철 생성 API 작성

Commit:
test: 인수 테스트 격리를 위한 cleanUp util 생성

- `dataSource` 활용 테이블 정보 획득 후 `truncate`

Commit:
feat: 지하철 노선 목록 조회 API 구현

- 노선 목록 조회 인수테스트 생성

Commit:
feat: 지하철 노선 조회 API 구현

- 노선 조회 인수테스트 생성
- 노선 조회 API 생성

Commit:
feat: 지하철 노선 수정 API 구현

- 노선 수정 인수테스트 생성
- 노선 수정 API 생성

Commit:
feat: 지하철 노선 삭제 API 구현

- 노선 삭제 인수테스트 생성
- 노선 삭제 API 생성

yunbaek committed:
refactor: 인수 테스트 공통 설정 추출

shared-moon commented:
안녕하세요!
질문 주신 내용은 여기에 따로 코멘트 할게요 :)

1. CleanUp 서비스 만드는 건 좋은 방법이라고 생각합니다! 컨텍스트를 따로 띄우는건 테스트 비용이 너무 많이 들어서 개인적으론 사용하지 않는 방법이에요 :) 참고로 저희 팀에서는 Repository테스트시엔 `DbUnit`을 사용하고 있습니다 !

2. 나중에 TDD를 이해하시면 조금 명확해 지실 것 같은데요, 테스트는 `어디까지 해야하나` 보다는 `무엇을 검증하려 하나`에 초점을 맞추시면 더 이해가 되지 않을까 싶네요! 말씀하신 걸 예로 들면, `조회에 대한 응답 검증` 이라는 건 결국 `조회가 잘 되는가` 이기때문에 id만 확인하셔도 충분 할 것 같습니다. 대신 `업데이트가 잘 되었나` 를 확인하려면 업데이트 대상 컬럼들이 잘 변경 되었는지를 봐야겠죠 :)

참고로 몇자 더 적어보자면.. 보통 라이브러리나 더 하위 레이어에서 이미 검증된 작업은 상위 레이어에서 굳이 더 검증하실 필요는 없습니다. 예를들면 Controller - Service 구조가 있을 때, Service에서 비즈니스 로직들이 검증이 되었다면, Controller에서는 굳이 비즈니스 로직에 대한 검증은 안하는 것 처럼요!

마찬가지로 조회에 대한 검증이라면 `findById()` 등일텐데, 커스텀한 메서드가 아니라면 굳이 검증하실 필요는 없다고 생각합니다 :) 저도 라이브러리 등은 따로 검증하진 않고 그냥 레퍼런스를 믿고 사용해요!

shared-moon commented on src/main/java/nextstep/subway/application/line/LineService.java:
```diff
@@ -0,0 +1,65 @@
+import nextstep.subway.exception.LineNotFoundException;
+
+@Service
+@Transactional(readOnly = true)
```
`@Transactional`을 클래스 단위에 선언하는건 왠만하면 지양하는게 좋을 것 같아요! 혹시 다르게 처리되어야 하는 메서드에 실수로 선언을 안하면 비정상적인 동작을 할 수가 있는데, 이게 클래스에 선언되어있으면 잘 안보여서요 :)

yunbaek replied:
해당 부분은 큰 고민 없이 습관적으로 붙였는데, 피드백 주신 부분 고민해보니, 사실 저도 같은 실수를 한 적이 있었던 것 같습니다 😂 
`(reeadOnly = true)` 옵션을 주었다고 `C, U, D` 동작에만 `@Transactional` 어노테이션을 붙이는 것을 지양하고 `@Transactional` 이 필요한 모든 메서드에 명시적으로 붙여보겠습니다! 

(방금 피드백 반영 하면서, 작다면 작은 서비스임에도 클래스단에서 선언 후 메서드에 선언해 주지 않은  `@Transactional(ReadOnly=true)` 를 당장 어디에 붙여야 될 지 햇갈렸습니다. 조언 주신대로 메서드마다 명시적으로 선언해주니 명확하게 파악하기도 쉬워지네요. 피드백 감사드립니다! 😄)

shared-moon commented on src/main/java/nextstep/subway/application/line/LineService.java:
```diff
@@ -0,0 +1,65 @@
+		Station upStation = stationService.findById(request.getUpStationId());
+		Station downStation = stationService.findById(request.getDownStationId());
+		Line line = lineRepository.save(
+			Line.of(request.getName(), request.getColor(), upStation, downStation, request.getDistance()));
```
도메인 객체가 상위 레이어에 의존성이 있는건 좋은 방법이 아닐 것 같아요!

yunbaek replied:
말씀대로 `DTO - DOMAIN` 간 의존성이 여기저기 있었습니다!
최대한 제거하려고 노력 해 보았는데, 올바르게 하고 있는 건지 걱정이 되네요:joy:
그리고 DTO 관련한 질문도 있어서 [아래](https://github.com/next-step/atdd-subway-admin/pull/914#discussion_r1023407804) 에 남겨 보았습니다!

shared-moon commented on src/main/java/nextstep/subway/application/line/LineService.java:
```diff
@@ -0,0 +1,65 @@
+		return LineResponse.of(line);
+	}
+
+	public List<LineResponse> findLines() {
```
여기도 findAll이라고 표현하는 건 어떨까요 ?

yunbaek replied:
말씀주신대로 `findAll` 이라고 짓는 것이 조금 더 명확할 것 같습니다!
메서드 작명에 신경 쓰겠습니다! 😄

shared-moon commented on src/main/java/nextstep/subway/application/line/LineService.java:
```diff
@@ -0,0 +1,65 @@
+	@Transactional
+	public void updateLine(Long id, LineUpdateRequest request) {
+		Line line = findById(id);
+		line.updateLine(request);
```
여기도 도메인이 DTO를 참조하고 있네요, 파라미터를 뭉쳐서 가져가고 싶으신 마음은 이해합니다만, 하위레이어는 상위레이어에 의존성이 없어야 결합도를 낮추고 재사용성을 늘릴 수 있어요!

yunbaek replied:
넵. 파라미터를 뭉쳐서 들고 오고 싶었는데, 다시 살펴보니 말씀대로 `Presentation DTO` 와 `domain` 을 분리하지 못했네요 😂 
피드백 주신 부분을 고민해보면서 DTO 에 관한 여러가지 의견들도 살펴보았는데  추가적으로 고민되는 부분들이 있었습니다. 🤔 

1. 서비스가 `Presentation DTO` 를 알아도 될까?
    - `service`에 `Presentation DTO` 가 들어가게 되면 해당 서비스의 재사용이 어렵다.
    - `service`가 `Presentation DTO` 를 사용하지 않게 되면 로직에 따라 `Controller` 가 너무 많은 역할을 하게 될 수도 있다.

2. DTO 와 Entity 를 분리하면서 파라미터의 개수가 너무 많아지게 되면 어떻게 해야 할까?
    - 당장 생각나는 방식은 다시 묶어주는 방법인데, `VO` 의 개념으로 사용하는 것이 적절할까?
    - 만약 `VO` 를 통해 묶은 객체가 `DTO` 와 동일한 필드와 동일한 값(isSameAS) 을 갖게 되더라도 분리하는 것이 적절할까? 

기초적인 지식이 없어서 그런지 공부를 하면 할 수록 정리가 되는 것이 아니라 고민만 늘어나는 것 같습니다..ㅠㅠ
너무 추상적인 질문인 것 같아 죄송하지만, 기준(?) 을 잡는데에 조금만 도음을 주셨으면 좋겠습니다. 🙏 
항상 좋은 필드백 감사드립니다!

shared-moon replied:
1. 어디까지 격리를 할 것인가의 문제라고 생각하는데요, 일단 말씀하신 Presentation DTO는 api Request DTO로 이해를 했습니다! controller와 service를 한 그룹으로 본다면 request를 그대로 사용해도 될 것 같구요, 대신 도메인 레이어는 두 단계를 넘어가는거라 거기까지 전파되는 건 좋지 않을 것 같습니다. 중요한 건 의존성 방향이 반대로 가지 않게 하는 것일 것 같아요! 도메인이 request에 의존성을 가지게 돼 버리면 그 도메인은 해당 컨트롤러를 통해서만 사용이 가능해져 버리기 때문에 결합도가 증가하니까요 :)

만약 controller와 service를 격리하고 service의 재사용성을 높이고 싶으시다면, service 앞에 facade 등으로 전처리를 위한 녀석을 더 둘 수도 있습니다! 대신 이런 경우 service는 본인이 소화 가능한 request와 response를 정의해야해서, 각 단계별로 DTO가 꽤 많이 늘어날 수 있어요! 어떤 비즈니스인지, 서비스가 얼마나 재사용 될지에 따라 선택 해야 할 문제일 것 같아요!

2. Entity에 넘어가는 파라미터가 많다는 건, Entity에 그만한 기능이 생긴다는 뜻이겠죠 ? 그렇다면 많은 파라미터를 넘기는 걸 고민하기 전에, 메서드의 책임이 너무 과중하진 않은가 ? 설계가 잘못되진 않았나 ? 를 먼저 고민 해 봐야 할 것 같습니다! 일반적인 상황에서 도메인 레이어의 비즈니스 로직이 많은 파라미터를 받는 경우는 잘 없을 것 같아요!

shared-moon commented on src/main/java/nextstep/subway/dto/line/LineCreateRequest.java:
```diff
@@ -0,0 +1,45 @@
+		this.distance = distance;
+	}
+
+	public static LineCreateRequest of(String name, String color, Long upStationId, Long downStationId, int distance) {
```
of 메서드가 별로 효용이 없어보이네요! 꼭 필요한 상황일까요 ?

yunbaek replied:
음..단순 DTO 를 생성하는데에 `정적 팩토리 메서드` 사용은 오버엔지니어링이 될 수 있다는 말씀이실까요?? 🤔 
`Effective java` 를 막 읽기 시작해서, 제일 처음에 나온 `생성자 보다는 정적 팩토리 메서드를 사용하라` 에 너무 심취했었나 보네요 😂

shared-moon replied:
정적 팩토리 메서드를 권장하는 이유에 대해서 다시 한 번 생각 해 보시면, 지금 상황에선 왜 굳이 필요가 없을지 조금 더 고민 해 볼 수 있을 것 같아요!
자바에서의 생성자가 비효율적인 이유는

1. 네임드 파라미터를 지원 안한다
2. 필드가 여러 개 존재 할때, allArgumentsConstructor 외의 다른 생성자가 존재할 수 있고, 동작이 다를 수 있다

뭐 이런 이유들 때문인데, of 메서드가 효과를 보려면 다른 DTO를 받아서 컨버팅을 하는 형태가 되어야 할 것 같아요! 아니면 빌더패턴을 사용하는건데 Lombok의 지원이 없으면 리소스 낭비가 좀 심해서 ..

지금 상황은 of가 생성자와 완벽하게 동일한 역할을 수행하고 있기 때문에 위에 있던 문제가 전혀 개선되지 않아 보여서 의미가 있을까? 하는 질문을 드려봤어요 :)

shared-moon commented on src/test/java/nextstep/subway/util/DatabaseCleanUpUtils.java:
```diff
@@ -0,0 +1,54 @@
+	@Transactional
+	public void cleanUp() {
+		try {
+			Connection connection = dataSource.getConnection();
```
혹시 테스트 패키지 전체로 돌리면 중간에 안돌아가지 않나요 ?
커넥션 누수에 대해 한번 알아보시면 좋을 것 같아요!

yunbaek replied:
헉..전체 테스트를 돌리는 걸 깜빡해서 해당 오류를 몰랐었네요 😂 
커넥션을 열어놓고 닫는 것을 깜빡 했습니다 😢
수정하겠습니다!

shared-moon reviewed (changes requested):
경윤님 안녕하세요!
2단계 미션 잘 진행 해 주셨네요 :)
코멘트를 몇개 남겨뒀는데요, 이번 미션도 한번 짚어보면 좋을 것 같아서 확인 후 리뷰 재요청 부탁드릴게요!
화이팅입니다 🔥

Commit:
refactor: Step2 피드백 반영(1)

- `DTO` - `entity` 분리
- CleanUp 에 사용한 Connection Close
- 메서드 명 변경
- `@Transactional` 어노테이션의 명확한 사용
- 불필요한 정적 팩토리 메서드 제거

Commit:
refactor: Step2 피드백 반영(2)

- `DTO` - `Domain` 간 의존성 제거

Commit:
refactor: Step2 피드백 반영(3)

- `DTO` - `Domain` 간 의존성 제거 추가

shared-moon commented on src/test/java/nextstep/subway/util/DatabaseCleanUpUtils.java:
```diff
@@ -47,6 +47,7 @@ public void cleanUp() {
 				connection.createStatement().execute("ALTER TABLE " + tableName + " ALTER COLUMN ID RESTART WITH 1");
 			}
 			connection.createStatement().execute("SET REFERENTIAL_INTEGRITY TRUE");
+			connection.close();
```
위에 statement도 계속 생성되고 있는데, createStatement() 를 로컬변수로 받고 재사용 할 수 있을 것 같아요! 그리고 얘도 쓴 다음에 닫아줘야합니다!

shared-moon reviewed (approved):
경윤님 안녕하세요!
피드백 반영 잘 해 주셨네요 :)
질문 주신 내용의 답변과 코멘트를 좀 남겨 보았는데요
다음 미션 진행 전에 확인 부탁드립니다!
고생 많으셨어요 :)
